scripts/generate-abis.ts

```diff
@@ -0,0 +1,68 @@
+import * as path from 'path';
+import * as fs from 'fs';
+
+/**
+ * This generates const TypeScript ABIs for each asset in
+ * `@safe-global/safe-deployments` package for `viem` to infer.
+ *
+ * Although it is possible to get a singleton programmatically and
+ * import the JSON directly, neither is strictly typed.
+ *
+ * Once it is possible to import JSON "as const", the deployments
+ * package should be updated to return the singletons as such.
+ *
+ * @see https://github.com/microsoft/TypeScript/issues/32063
+ */
+
+// Path to directory containing JSON assets
+const assetsDir = path.join(
+  process.cwd(),
+  'node_modules',
+  '@safe-global',
+  'safe-deployments',
+  'dist',
+  'assets',
+);
+
+// Path to directory where ABIs will be written
+const outputDir = path.join(process.cwd(), 'src', 'abis', 'safe');
+
+function main() {
+  // Remove any existing ABIs
+  try {
+    fs.rmSync(outputDir, { recursive: true });
+  } catch {
+    // Swallow error if directory does not exist
+  }
+
+  // For each version...
+  for (const version of fs.readdirSync(assetsDir)) {
+    const versionOutputDir = path.join(outputDir, version);
+
+    fs.mkdirSync(versionOutputDir, { recursive: true });
+
+    const versionDir = path.join(assetsDir, version);
+
+    // ...parse the ABI for each asset
+    for (const assetFile of fs.readdirSync(versionDir)) {
+      // Read the asset JSON
+      const assetPath = path.join(assetsDir, version, assetFile);
+      const assetJson = fs.readFileSync(assetPath, 'utf8');
+
+      // Parse the asset JSON
+      const { contractName, abi } = JSON.parse(assetJson);
```

Comment:
This can be a bottleneck if the JSON file is large. We should look into working with the Stream API if possible – https://nodejs.org/api/stream.html#stream

However reading and validating the JSON file might require the complete context and might be tricky to process it in chunks without relying on an external tool. Currently we would be holding the entire JSON file in memory (the `assetJson`).

We can potentially use `createWriteStream` with this solution:

```typescript
const stream = fs.createWriteStream(filePath);
stream.write(
  '// This file is auto-generated by scripts/generate-abis.ts\nexport default',
);
stream.write(data);
stream.write('as const;');
stream.end();
```

But we wouldn't get a lot of benefits out of it given that the stream would be used to write the entire "chunk" (i.e.: the entire file) that is in memory (instead of having a `pipeline` to process chunks of it).

Given the complexity of validating the JSON file in chunks I would say that this solution is totally fine as it's also a step of the build process and not a service runtime operation.

Reply:
As you said, I don't think there will be much benefit from using the Steam API as this doesn't run at runtime. However, I altered to using it in 9dff79a72bbe821c2a4bb3861949c01770d340c4 as it gives a clear separation of: comment, data and const assertion. I've added a comment summarising your points for future reference as well.